FBP creation

Eagle0600 opened the pull request:
Allows creation of FBPs by adding a robotic head to a robotic torso.
Heads and torsos respect brand.
Robot analyzer readout no longer reports "system instability", as the function used to get that number will always return 0 for any valid target.

Eagle0600 committed:
Removed "system instability" readout from robot analyzer.

Synthetic humans are defined as being immune to toxloss under
"/mob/living/carbon/human/getToxLoss()" in the file
"/code/modules/mob/living/carbon/human/human_damage.dm".

Eagle0600 committed:
Revert "Removed "system instability" readout from robot analyzer."

This reverts commit b459d0bc0373df6ea032108cd4c781ff7fb28965.

Eagle0600 committed:
Revert "Revert "Removed "system instability" readout from robot analyzer.""

This reverts commit 713437358ce9b8bd6803b6abfc6370ba33cf4a42.

Eagle0600 committed:
Changelog

"Removed 'system instability' from robot analyzer."

Eagle0600 committed:
Fix of previous.

Eagle0600 committed:
Allow creation of full-body prosthetics.

Add a robot head to a robot torso. Respects brand (but torsos and heads
can't have brands).
Resulting FBP has sensors (no need to add) but no brain.

Eagle0600 committed:
Changelog for FBP creation.

Eagle0600 committed:
Allow brand-specific robotic heads and torsos.

This is so you can customise new FBPs.

Eagle0600 committed:
Changelog for robotic head and torso brands.

Eagle0600 commented:
This feature cannot be fully tested without at least one other player, but creation of a new FBP appears to work.

Planned features: Naming your new FPB (is this done automatically when the MMI is installed?), gender and species selection.

DeeUnderscore commented:
From testing this, MMI transplantation works, and no, it doesn't cause name rewrite.

Eagle0600 commented:
Well, that's great news. That it works, that is. As for the name, adding a name prompt should be easy.

Do you think it would be better to add it when the FBP is first created or when the MMI is put in?

atlantiscze commented:
For testing, i have found following method acceptable, but still a bit impractical. It may still be of use to you, however:
1. Start the server and join with your account as an admin.
2. Enable guests via the relevant admin verb.
3. Switch to BYOND, log out. That will turn you into the guest. Do not close the game window. You will stay logged in there.
4. Join your server again. This time you will join as a guest, that can be used as test victim.

Eagle0600 commented:
Thanks.

GinjaNinja32 commented:
You can also VV your client's key to something else then rejoin with the same account; BYOND won't kick the first connection because the key is different.

Eagle0600 commented:
Oh, I think I'd rather do that.

Eagle0600 committed:
Allows the roboticist to select a name and species for the new FBP.

Eagle0600 committed:
Changelog for previous commit.

Eagle0600 commented:
I can now confirm transfer works properly. Nonhuman species look a bit weird, but I think that's just how the robotics system works.

Kelenius commented on html/changelogs/Eagle0600-dev-2.yml:
```diff
@@ -0,0 +1,36 @@
+#################################
+
+# Your name.  
+author: N3X15
```
Woop.

Eagle0600 replied:
Thanks for catching that.

Eagle0600 committed:
Fixed changelog author.

Eagle0600 commented:
Alright, that should fix the changelog oversight on my part.

Eagle0600 commented:
Are there any other problems with this PR that need to be addressed?

PsiOmegaDelta commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -267,6 +269,64 @@
+		if(!src.cell)
+			user << "<span class='warning'>You need to attach a cell to it first!</span>"
+			success = 0
+		if(!(W:flash2 && W:flash1))
```
Colons. 
Use `/obj/item/robot_parts/head/ph = W` then `ph.flash1/2`.

Eagle0600 replied:
Again, this was copied from the preceding code. Should that be changed too?

Eagle0600 replied:
Done.

PsiOmegaDelta replied:
...this is copy-paste?

Eagle0600 replied:
The test in question was. From the test for adding a robot head to an endoskeleton to make a cyborg.

PsiOmegaDelta replied:
If the same test is in two places I'd rather see it become a proc, i.e.

```
var/obj/item/robot_parts/head/proc/can_install(var/mob/installer)
    if(flash1 && flash2)
        return TRUE
    else
        to_chat(installer, "<span class='warning'>You need to attach a flash to it first!</span>")
```

Then instead of

```
if(!(head_part.flash2 && head_part.flash1))
    user << "<span class='warning'>You need to attach a flash to it first!</span>"
    success = 0
```

You can do

```
success = head_part.can_install(user) && success
```

Repeat the check for the other location. The check is now centralized.

Eagle0600 replied:
Alright, I'll get on that.

Eagle0600 replied:
This is turning into a larger clean-up than I expected. It's a little outside of the scope of my PR, but do you mind if I do a general clean-up of robot_parts handling to reduce code duplication?

Eagle0600 replied:
Alright, that's done. It didn't affect as much stuff outside of robot_parts as it could have. Just about two to three sections of code that weren't specifically to do with robot construction.

PsiOmegaDelta commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -267,6 +269,64 @@
+
+			// Create a new, nonliving human.
+			var/mob/living/carbon/human/H = new /mob/living/carbon/human(loc)
+			H.stat = DEAD
```
Call `H.death(0, "no message")` or H will still remain in the living mob list, among other consistency issues.

Eagle0600 replied:
Done.

PsiOmegaDelta commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -267,6 +269,64 @@
+				species = "Human"
+			var/name = sanitizeSafe(input(user,"Set a name for the new prosthetic."), MAX_NAME_LEN)
+			if(!name)
+				name = "prosthetic ([rand(0,999)])"
```
Use `random_id(unique_key, min_number, max_number)`.
Example: `random_id(/obj/item/robot_parts, 1, 999)`

Alternatively: `sequential_id(/obj/item/robot_parts)`

Eagle0600 replied:
This was copied from the cloning code. Does that need changes too?

Eagle0600 replied:
Done.

PsiOmegaDelta replied:
If you can see a way to cleanly share the code through a proc, as above, then please do.

Eagle0600 replied:
I'll take a look.

Eagle0600 replied:
Nothing that would actually significantly reduce code reuse. You'd just be replacing `"clone ([rand(0,999)])"` with something like `unique_name("clone", 0, 999)`.

Eagle0600 commented:
Alright, compiling and running those changes to make sure it still works as-intended.

Eagle0600 committed:
Code cleanup

Eagle0600 commented:
Requested changes have been made and tested.

NullSnapshot commented:
I feel this is something that should probably be put up to community discussion. Just my two cents as a developer.

Eagle0600 committed:
robot_parts code cleanup

Consolidation of similar code
Changed the way robot_suit stores component parts when creating a cyborg
(so that they may be tested for similarly).

PsiOmegaDelta commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -272,16 +253,9 @@
-			success = 0
+		var/success = TRUE
+		success &= can_install(user)
+		success &= head_part.can_install(user)
```
This is the briefer method, however unlike before no more feedback is given after the first failure as this is equivalent to:

```
success = success && can_install(user)
success = success && head_part.can_install(user) // head.can_install won't be called if success if false.
```

If it's desired to inform the user of all issues at the same time you'll manually have to write.

```
success = can_install(user) && success 
success = head_part.can_install(user) && success 
```

GinjaNinja32 replied:
this is an `&=`, there is no short-circuiting `&&=`; this will still print all the feedback;

``` dm
/proc/foo()
    world.log << "foo"
    return 1

/world/New()
    var/x = 0
    x = x & foo() // Prints all feedback because & does not short-circuit
    x &= foo() // Same as above
    x = x && foo() // Does not print all feedback because && short-circuits
    x = foo() && x // Prints all feedback, as foo() comes first
    x &&= foo() // Doesn't compile, there is no &&= operator
```

PsiOmegaDelta replied:
Hm. Might add that to my toolbox of tricks then.

Eagle0600 replied:
Avoiding short-circuiting was a design consideration for these statements. Thanks for noticing that, by the way; it was entirely possible that I was wrong.

PsiOmegaDelta commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -125,55 +145,15 @@
 			user.drop_item()
-			W.loc = src
-			src.head = W
+			part.loc = src
```
part.forceMove(src)

PsiOmegaDelta commented:
Community feedback is probably good, codewise I have no major complaints.

Eagle0600 commented:
Well, how would I go about getting community feedback?

Kasuobes commented:
https://baystation12.net/forums/view/s-f/

Eagle0600 commented:
Created discussion thread [here](https://baystation12.net/forums/threads/discussion-full-body-prosthetic-creation.2728/).

Eagle0600 commented:
So far, four have solidly weighed in _for_ (not including myself), and one solidly _against_. Discussion has started to slow down.
